iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023, Eclipse Foundation and the iceoryx contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_HOOFS_POSIX_IPC_UNIX_DOMAIN_SOCKET_INL
+#define IOX_HOOFS_POSIX_IPC_UNIX_DOMAIN_SOCKET_INL
+
+#include "iox/posix_call.hpp"
+#include "iox/unix_domain_socket.hpp"
+
+namespace iox
+{
+template <typename Type>
+expected<void, PosixIpcChannelError> UnixDomainSocket::sendImpl(const Type* msg,
```

Comment:
I got some issues using ```not_null```.
This is a particular case where the functions are templates: the compiler does not deduce the type. Do you have any suggestions?

```
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp: In instantiation of ‘iox::UnixDomainSocket::timedReceive(const iox::units::Duration&) const::<lambda(auto:8*, auto:9)> [with auto:8 = char; auto:9 = iox::BufferInfo; uint64_t = long unsigned int]’:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/functional/include/iox/detail/function_ref.inl:45:90:   required from ‘iox::function_ref<ReturnType(ArgTypes ...)>::function_ref(CallableType&&) [with CallableType = iox::UnixDomainSocket::timedReceive(const iox::units::Duration&) const::<lambda(auto:8*, auto:9)>; <template-parameter-2-2> = void; ReturnType = long unsigned int; ArgTypes = {char*, iox::BufferInfo}]’
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:308:6:   required from here
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:302:16: error: no matching function for call to ‘iox::UnixDomainSocket::timedReceiveImpl(char*&, const uint64_t&, const iox::units::Duration&, iox::UnixDomainSocket::Termination) const’
  302 |         result = this->timedReceiveImpl(str, info.total_size, timeout, Termination::NULL_TERMINATOR);
In file included from /home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:19:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/unix_domain_socket.hpp:163:46: note: candidate: ‘template<class Type> iox::expected<long unsigned int, iox::PosixIpcChannelError> iox::UnixDomainSocket::timedReceiveImpl(iox::not_null<Type*>, uint64_t, const iox::units::Duration&, iox::UnixDomainSocket::Termination) const’
  163 |     expected<uint64_t, PosixIpcChannelError> timedReceiveImpl(not_null<Type*> msg,
      |                                              ^~~~~~~~~~~~~~~~
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/unix_domain_socket.hpp:163:46: note:   template argument deduction/substitution failed:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:302:16: note:   mismatched types ‘iox::not_null<Type*>’ and ‘char*’
  302 |         result = this->timedReceiveImpl(str, info.total_size, timeout, Termination::NULL_TERMINATOR);
```

Reply:
Oh, you have to use `not_null<Type>` instead of `not_null<Type*>`. At least with clang it worked when passing a `not_null(str)` to the method.

Reply:
If I remember correctly, I tried both signatures, but the compilation failed.
This is the first case in the entire code that ```not_null``` is used with a  template parameter. 
I'll check better, but I'm pretty sure about that.

Reply:
Sorry, you are right. It can be `not_null<Type*>`. I guess `Type` being a template parameter somehow prevents the implicit conversion to `not_null`. By explicitly using the ctor the problem is gone :)

Reply:
Specified the template parameter in the function call the code works